lib/tlscontext.c

```diff
@@ -305,6 +305,33 @@ file_exists(const gchar *fname)
   return TRUE;
 }
 
+static gboolean
+tls_context_setup_ecdh(SSL_CTX *ctx, const char *ecdh_curve_name)
+{
+  int ecdh_nid = OBJ_sn2nid(ecdh_curve_name);
+  if (ecdh_nid == NID_undef)
+    return FALSE;
+
+#if OPENSSL_VERSION_NUMBER > 0x10100000L
+  /* No need to setup as ECDH auto is the default */
+#elif OPENSSL_VERSION_NUMBER > 0x10002000L
+  if (SSL_CTX_set_ecdh_auto(ctx, 1) != 1)
+    return FALSE;
+#elif OPENSSL_VERSION_NUMBER > 0x10001000L
+  EC_KEY *ecdh = EC_KEY_new_by_curve_name(ecdh_nid);
+  if (!ecdh)
+    return FALSE;
+
+  if (SSL_CTX_set_tmp_ecdh(ctx, ecdh) != 1)
+    return FALSE;
+  EC_KEY_free(ecdh);
+#else
+  return FALSE;
```

Comment:
Can't we somehow extract this function to compat/openssl-support? That's where we add upgrade code to old openssl so we can use the latest API in our own code.

Would be great.

Hmm what if we defined SSL_CTX_set_ecdh_auto in there, doing roughly the same that you have here, except for cases where openssl already defines it.